### PR TITLE
ChainDB: optimise chain selection for forks

### LINKED ISSUE #1223: Optimize chain selection in ChainDB
The chain selection code in the chain DB follows the specification very closely, but of course the specification is not concerned with performance, only efficiency.  Chain selection works by finding a set of candidates starting at a particular point; if a new block comes in, and that block fits onto our current chain, the set of candidates is computed from current tip of the chain. However, if that block does _not_ find onto our chain, we construct a prefix path backwards from that block to the tip of the immutable DB, find all candidates starting at the new block, and then construct the set of candidates by prepending that prefix path to those candidates.

```
imm DB tip I-------|-------------------------- our chain
                   |
                   --------------B new header

imm DB tip I=======X-------------------------- our chain
                   |
                   |   rollback
                   ===============B----------
                                  |
                                  *-------------------
                                  |
                                  *--------------                   
```

We then construct a set of "candidate suffixes", by computing an intersection of that chain with our current chain, and then using that to compute a rollback (in terms of a number of blocks) and a set of new headers. However, of course, the rollback will in fact be the _same_ for all of those candidates, and there was no need to recompute the intersection point at all, because we knew that as soon as we computed the prefix back from the new header to our current chain. 

This means that the implementation is _correct_, but is doing unnecessary work, which will become important once chains are more frequent. Therefore marking this as high priority, but only for shelley.

### PR DESCRIPTION
**Pull request by mrBliss**
Fixes #1223.

When a block fits onto a fork, we construct the fragment all the way going
back to the immutable tip instead of just a `ChainDiff` to the most recent
intersection with the current chain. For example, when switching to a fork
that requires rolling back 2 blocks, previously, we'd construct a fragment
containing `k` headers. Now, we'll construct a `ChainDiff` with rollback = 2
and a fragment of 2 headers.

* `VolDB`: introduce `ReversePath` and `computeReversePath` which lazily
  computes a path through the VolatileDB. By using these paths everywhere, we
  no longer have to think about looking things up in the VolatileDB.

* `VolDB`: let `isReachable` return a `ChainDiff` of `HeaderFields` instead of
  a list of hashes going all the way back to the immutable tip. This
  `ChainDiff` of `HeaderFields` can later be translated to a `ChainDiff` of
  `Header`s.

* `VolDB`: add `extendWithSuccessors` that will extend a `ChainDiff` of
  `HeaderFields` with its successors, using `candidates`.

* Add property tests for `isReachable`, which helped catch a tricky corner
  case involving EBBs.

* `VolDB`: rewrite `computePath` using `computeReversePath`.

* We can now use `RealPoint`s instead of `HeaderHash`s in many functions and
  types: `IteratorBlockGCed`, `ChainDB.Iterator`s, `VolDB.Path`, some trace
  events, etc.

We don't yet take advantage of the `RealPoint` in the implementation of
`ChainDB.Iterator`s.

### CODE REVIEW

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs [hunk 1]
```diff
@@ -268,47 +273,221 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
...
+{-------------------------------------------------------------------------------
+  Reverse paths
+-------------------------------------------------------------------------------}
+
+-- | Return the block info for the block with the given hash. Return 'Nothing'
+-- when not in the VolatileDB.
+type LookupBlockInfo blk =
+  HeaderHash blk -> Maybe (VolDB.BlockInfo (HeaderHash blk))
+
+data ReversePath blk =
```
[the remark this answers is not in the document]
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:285]: Docstring: a path through the VolatileDB starting at a block in the VolatileDB until we reach genesis or "leave" the VolatileDB

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs [hunk 2]
```diff
@@ -268,47 +273,221 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
...
-        NotInVolDB        _      ->
-          error "impossible: block just added to VolatileDB missing"
+-- A 'PointChainDiff' is returned iff @P@ is on the chain fragment. Moreover,
+-- when the number of blocks to roll back is also 0, it must be that @P@ is
+-- the tip of the chain fragment.
+--
+-- When the returned list of points is non-empty, @P@ will be the last point
+-- in the suffix of the 'PointChainDiff'.
+isReachable
+  :: forall blk. ((HasHeader (Header blk)), HasCallStack)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:388]:
```suggestion
  :: forall blk. (HasHeader (Header blk), HasCallStack)
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs [hunk 3]
```diff
@@ -268,47 +273,221 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
...
+  => LookupBlockInfo blk
+  -> AnchoredFragment (Header blk)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:390]: Comment: current chain

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs [hunk 4]
```diff
@@ -268,47 +273,221 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
...
+      Nothing          -> Nothing
+      Just reversePath -> go chain reversePath 0 []
   where
-    from = StreamFromExclusive i
-    to   = StreamToInclusive   b
+    -- | NOTE: the 'ReversePath' is lazy in its spine. We will only force as
+    -- many elements as 'RealPoint's we return. In the worst case, the path is
+    -- not connected to the current chain at all, in which case we do force
+    -- the entire path.
+    go
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:403]: Document: we're trying to find a common block, i.e., one with the same point and thus the same slot. Both the chain and the path are ordered by slots, so we compare the slots and drop the largest one until we have a match in slot, then we check hashes. If those don't match, we drop both.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs [hunk 5]
```diff
@@ -268,47 +273,221 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
...
+      StoppedAtGenesis
+
+      -- | The reverse path stopped at this hash, which is the hash of the
+      -- predecessor of the last block in the path (that was still stored in
+      -- the VolatileDB).
+      --
+      -- The block corresponding to the predecessor is /not/ stored in the
+      -- VolatileDB. Either because it is missing, or because it is old and
+      -- has been garbage collected.
+    | StoppedAt !(HeaderHash blk)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:296]: Remove bangs from this type

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs [hunk 6]
```diff
@@ -268,47 +273,221 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
...
+
+      -- | Snoc: the given 'RealPoint' is in the VolatileDB.
+      --
+      -- NOTE: we are intentionally lazy in the spine, as constructing the
+      -- path requires lookups in the VolatileDB's in-memory indices, which
+      -- are logarithmic in the size of the index.
+    | (ReversePath blk) ::> !(RealPoint blk)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:303]: Include `IsEBB`!

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs [hunk 7]
```diff
@@ -558,37 +554,43 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr = do
...
+                  )
+              . catMaybes
+              )
+          $ flip evalStateT initCache
+          $ mapM (constructChainDiff pdiff)
+          $ if null suffixesAfterB
+              -- If there are no suffixes after @b@, just use the diff that
+              -- ends in @b@ as the sole candidate.
+            then [[]]
+            else map NE.toList suffixesAfterB
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs:593]: Later: extend the pdiffs first, then translate to diff.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs [hunk 8]
```diff
@@ -558,37 +554,43 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr = do
...
-                filter
-                  (preferAnchoredCandidate cdbTopLevelConfig curChain .
-                   Diff.getSuffix)
-              . mapMaybe (Diff.diff curChain)
-              $ NE.toList candidates
+            -- We use a cache to avoid reading the headers from disk multiple
+            -- times in case they're part of multiple forks that go through
+            -- @b@.
+            initCache      =  Map.singleton (headerHash hdr) hdr
+        chainDiffs <-
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs:577]: The five steps:

1. Compute candidates (`VolDB.candidates`) as `AnchoredFragment`s
2. Extend our diffs with them
3. Filter on suffix >= rollback
4. Translate diffs from `HeaderFields` to headers
5. Filter on `preferAnchoredCandidate`

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 9]
```diff
@@ -0,0 +1,340 @@
...
+    }
+
+deriving instance (HasHeader blk, Show (Header blk)) => Show (ReachableSetup blk)
+
+-- | Newtype that uses 'ppShow' for its 'Show' instance.
+newtype Pretty a = Pretty a
+  deriving (Eq, Arbitrary)
+
+instance Show a => Show (Pretty a) where
+  show (Pretty a) = ppShow a
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:75]: Get rid of this

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+    let initFragment = AF.Empty AF.AnchorGenesis AF.:> getHeader firstBlk
+    fullChain <- genChainHelper
+                   (return currentChainBody)
+                   size
+                   firstBlk
+                   initFragment
+    let mbAnchorHdr = case AF.head (AF.dropNewest takeNewestN fullChain) of
+                        Left _          -> Nothing
+                        Right anchorHdr -> Just anchorHdr
+    return (AF.anchorNewest (fromIntegral takeNewestN) fullChain, mbAnchorHdr)
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:150]: TODO: never an empty chain anchored at genesis, i.e., drop the `firstBlk`. (case at the start)

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+-- | NOTE: the fork must be the minimal fork, i.e., it must not contain headers
+-- that are also part of the current chain.
+--
+-- NOTE: unlike 'ChainDiff's, 'PointChainDiff's can be shorter than the
+-- current chain.
+generateFork ::
+     AnchoredFragment (Header TestBlock)
+  -> Gen (PointChainDiff TestBlock, [Header TestBlock])
+generateFork chain = do
+    -- Roll back 0 or more headers and add 0 ore more headers
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:161]:
```suggestion
    -- Roll back 0 or more headers and add 0 or more headers
```

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+    let mbAnchorHdr = case AF.head (AF.dropNewest takeNewestN fullChain) of
+                        Left _          -> Nothing
+                        Right anchorHdr -> Just anchorHdr
+    return (AF.anchorNewest (fromIntegral takeNewestN) fullChain, mbAnchorHdr)
+
+-- | NOTE: the fork must be the minimal fork, i.e., it must not contain headers
+-- that are also part of the current chain.
+--
+-- NOTE: unlike 'ChainDiff's, 'PointChainDiff's can be shorter than the
+-- current chain.
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:156]: Later: become `ChainDiff`

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 13]
```diff
@@ -0,0 +1,340 @@
...
+    rollback <- choose (0, AF.length chain)
+    let mbPrevHdr = case AF.head (AF.dropNewest rollback chain) of
+          Left anchor -> fakeHdrFromAnchor (AF.castAnchor anchor)
+          Right hdr   -> Just hdr
+    case mbPrevHdr of
+      Just prevHdr -> generateForkFromHeader  rollback prevHdr
+      Nothing      -> generateForkFromGenesis rollback
+  where
+    -- | Given an anchor, return the header corresponding to it so we can use
+    -- that as to create the "previous block" when calling 'genChainHelper'.
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:171]:
```suggestion
    -- that to create the "previous block" when calling 'genChainHelper'.
```

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 14]
```diff
@@ -0,0 +1,340 @@
...
+    -- We only use the hash, slot number, and block number in that function
+    -- (indirectly), so we can simply fake the other fields.
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:173]: Mention chain length

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 15]
```diff
@@ -0,0 +1,340 @@
...
+        AF.AnchorGenesis       -> Nothing
+        AF.Anchor sno hash bno -> Just $ TestHeader' $ TestHeader {
+            thHash        = hash
+          , thSlotNo      = sno
+          , thBlockNo     = bno
+            -- Unused fields, but strict
+          , thPrevHash    = GenesisHash
+          , thBodyHash    = TestBodyHash 0
+          , thChainLength = ChainLength 0
+          , thIsEBB       = RegularBlock
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:187]: Make all of them lazy again and use `OnlyCheckIsWHNF`

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 16]
```diff
@@ -0,0 +1,340 @@
...
+genSuccessor prevBlk genBody = frequency
+    [ ( if prevIsEBB then 0 else 1
+      , mkNextEBB (const True) prevBlk
+          <$> chooseSlot (prevSlot + 1) (prevSlot + 2)
+          -- We don't care about EpochNo
+          <*> return 0
+          <*> genBody
+      )
+    , (3
+      , mkNextBlock prevBlk
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:107]: Change `mkNextEBB` and `mkNextBlock` to take `HeaderFields` and `ChainLength`, so that we don't have to do all the faking.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 17]
```diff
@@ -0,0 +1,340 @@
...
+    prevSlot  = blockSlot prevBlk
+    prevIsEBB = isJust $ headerIsEBB $ getHeader prevBlk
+
+    chooseSlot :: SlotNo -> SlotNo -> Gen SlotNo
+    chooseSlot (SlotNo start) (SlotNo end) = SlotNo <$> choose (start, end)
+
+genChainHelper ::
+     Gen TestBody
+  -> Int                                  -- ^ Number of headers to generate
+  -> TestBlock                            -- ^ Previous block
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:122]: Take `WithOrigin TestBlock` (or `WithOrigin HeaderFields`) and do the choice between `genFirstBlock`/`genSuccessor` here.

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 18]
```diff
@@ -0,0 +1,340 @@
...
+    fakeHdrFromAnchor = \case
+        AF.AnchorGenesis       -> Nothing
+        AF.Anchor sno hash bno -> Just $ TestHeader' $ TestHeader {
+            thHash        = hash
+          , thSlotNo      = sno
+          , thBlockNo     = bno
+            -- Unused fields, but strict
+          , thPrevHash    = GenesisHash
+          , thBodyHash    = TestBodyHash 0
+          , thChainLength = ChainLength 0
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:186]: Base it on the `Maybe (Header TestBlock)` returns by `generateChain`?

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 19]
```diff
@@ -0,0 +1,340 @@
...
+                      (fromIntegral rollback)
+                      (AF.castAnchor anchor)
+                      pts
+        return (pdiff, AF.toOldestFirst suffix)
+
+    generateForkFromGenesis ::
+         Int
+      -> Gen (PointChainDiff TestBlock, [Header TestBlock])
+    generateForkFromGenesis rollback = sized $ \size -> do
+        toAdd  <- choose (0, size)
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:217]:
```suggestion
        toAdd <- choose (0, size)
```

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 20]
```diff
@@ -0,0 +1,340 @@
...
+        if toAdd == 0 then
+          return (PointChainDiff (fromIntegral rollback) AF.AnchorGenesis [], [])
+        else do
+          firstBlk <- genFirstBlock genBody
+          suffix     <- genChainHelper
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:222]:
```suggestion
          firstBlk <- genFirstBlock genBody
          suffix   <- genChainHelper
```

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 21]
```diff
@@ -0,0 +1,340 @@
...
+        [ (3, generateConnectedFork    currentChain)
+        , (1, generateDisconnectedFork currentChain)
+        ]
+      let blockInfo = forkBlockInfo
+                   <> headersToBlockInfo mbAnchorHdr
+                   <> headersToBlockInfo (AF.toOldestFirst currentChain)
+      return ReachableSetup { currentChain, forkTip, fork, blockInfo }
+    where
+      generateConnectedFork currentChain = do
+          (fork, headersOnFork) <- generateFork currentChain
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:297]: Use `suchThatMap` instead of `discard`

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 22]
```diff
@@ -0,0 +1,340 @@
...
+
+      generateDisconnectedFork currentChain = do
+          disconnectedHeaders <- generateDisconnectedHeaders currentChain
+          forkTip <- elements (map headerRealPoint (NE.toList disconnectedHeaders))
+          return (
+              forkTip
+            , Nothing
+            , headersToBlockInfo disconnectedHeaders
+            )
+  shrink ReachableSetup { currentChain, forkTip, fork, blockInfo } =
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:312]: Blank line

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 23]
```diff
@@ -0,0 +1,340 @@
...
+                ReachableSetup {
+                    currentChain
+                  , forkTip   = forkTip'
+                  , fork      = Just $ PointChainDiff rollback anchor (reverse revSuffix')
+                  , blockInfo = Map.delete (realPointHash forkTip) blockInfo
+                  }
+              _ -> []
+
+      shrinkCurrentChain = case fork of
+          Nothing -> []
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:329]: Leave a comment: we don't care about this case

File: ouroboros-consensus/src/Ouroboros/Consensus/Fragment/ValidatedDiff.hs [hunk 24]
```diff
@@ -43,16 +47,18 @@ pattern ValidatedChainDiff d l <- UnsafeValidatedChainDiff d l
...
   -> l
-  -> ValidatedChainDiff blk l
+  -> ValidatedChainDiff b l
 new chainDiff ledger =
     assertWithMsg precondition $
     UnsafeValidatedChainDiff chainDiff ledger
   where
-    chainDiffTip = getTip chainDiff
+    chainDiffTip, ledgerTip :: Point b
+    chainDiffTip = Diff.getTip chainDiff
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Fragment/ValidatedDiff.hs:61]: Use `ledgerTipPoint'`?

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs [hunk 25]
```diff
@@ -264,47 +271,268 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
...
+-- | Return the block info for the block with the given hash. Return 'Nothing'
+-- when not in the VolatileDB.
+type LookupBlockInfo blk =
+  HeaderHash blk -> Maybe (VolDB.BlockInfo (HeaderHash blk))
+
+headerFieldsFromBlockInfo ::
+     VolDB.BlockInfo (HeaderHash blk)
+  -> HeaderFields blk
+headerFieldsFromBlockInfo VolDB.BlockInfo { bslot, bbid, bbno } =
+    HeaderFields {
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:322]: Try making the fields strict. If that works, use it in `VolDB.BlockInfo` and `Anchor`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs [hunk 26]
```diff
@@ -264,47 +271,268 @@ candidates succsOf b = mapMaybe NE.nonEmpty $ go (fromChainHash (pointHash b))
...
+    --
+    -- > block:         regular block 1 | EBB | regular block 2
+    -- > block number:                X |   X | X + 1
+    --
+    -- Edge case: if there are two or more consecutive EBBs, we might predict
+    -- the wrong block number, but there are no consecutive EBBs in practice
+    -- (nor in the tests), they are one epoch apart.
+    prevBlockNo :: BlockNo -> IsEBB -> BlockNo
+    prevBlockNo bno isEBB = case (bno, isEBB) of
+      (0, IsNotEBB) -> error "block with blockNo 0 cannot have a predecessor"
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:412]: Byron:
```
Chain:           Genesis :] Block :] Block
Block numbers:   Origin     At 0     At 1
Is EBB:          NA         yes      no
Has predecessor: NA         no       yes
```
Shelley:
```
Chain:           Genesis :] Block :] Block
Block numbers:   Origin     At 0     At 1
Is EBB:          NA         no       no
Has predecessor: NA         no       yes
```
Conclusion: a block with block number 0 is always the very first block and can't have a predecessor. However, that `IsNotEBB` must be removed from that pattern.

Add precondition: must have predecessor.

Rewrite the error message to: "precondition violated".

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs [hunk 27]
```diff
@@ -318,56 +546,90 @@ isReachable predecessor i b =
...
+        RealPoint (headerFieldSlot flds) (headerFieldHash flds)
+
+    -- | Convert the 'HeaderFields' to a 'RealPoint' and prepend that to the
+    -- accumulator.
+    --
+    -- NOTE: forces the 'RealPoint' to avoid holding onto the lazy
+    -- 'HeaderFields' data type, which is derived from a 'VolDB.BlockInfo'
+    -- (see 'computeReversePath'), which is much larger than a 'RealPoint'.
+    addToAcc :: HeaderFields blk -> [RealPoint blk] -> [RealPoint blk]
+    addToAcc flds = (pt :)
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:592]: Eta-expand and add comment: if this list is forced, that forces `pt`, and the list is forced because the accumulator is forced in `go`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs [hunk 28]
```diff
@@ -318,56 +546,90 @@ isReachable predecessor i b =
...
+    fieldsToRealPoint :: HeaderFields blk -> RealPoint blk
+    fieldsToRealPoint flds =
+        RealPoint (headerFieldSlot flds) (headerFieldHash flds)
+
+    -- | Convert the 'HeaderFields' to a 'RealPoint' and prepend that to the
+    -- accumulator.
+    --
+    -- NOTE: forces the 'RealPoint' to avoid holding onto the lazy
+    -- 'HeaderFields' data type, which is derived from a 'VolDB.BlockInfo'
+    -- (see 'computeReversePath'), which is much larger than a 'RealPoint'.
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:590]: Replace with: we'll store this in the state of an iterator, which is a lazy list.

See https://github.com/input-output-hk/ouroboros-network/issues/2341.

Run with the no unexpected thunks check.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs [hunk 29]
```diff
@@ -559,40 +555,47 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr = do
...
+              )
+            -- 3. Translate the 'HeaderFields' to 'Header' by reading the
+            -- headers from disk.
+          . flip evalStateT initCache
+          . mapM translateToHeaders
+            -- 2. Filter out candidates that are shorter than the current
+            -- chain. We don't want to needlessly read the headers from disk
+            -- for those candidates.
+          . NE.filter (not . Diff.isShorter)
+            -- 1. Extend the diffs with candidates fitting on @B@
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs:596]:
```suggestion
            -- 1. Extend the diff with candidates fitting on @B@
```

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs [hunk 30]
```diff
@@ -688,44 +690,32 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr = do
...
-
+    translateToHeaders (ChainDiff rollback flds) =
+        fmap
+          ( ChainDiff rollback
+          . AF.fromOldestFirst (AF.castAnchor (AF.anchor flds))
+          )
+      . mapM (getKnownHeaderThroughCache cdbVolDB)
+      . map headerFieldHash
+      . AF.toOldestFirst
+      $ flds
```
Comment (mrBliss) [inline ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs:718]: Add and use `ChainDiff.mapM`

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 31]
```diff
@@ -0,0 +1,326 @@
...
+      NotOrigin prevInfo@(prevHeaderFields, _, _) -> do
+        let anchor  = AF.Anchor
+                        (headerFieldSlot    prevHeaderFields)
+                        (headerFieldHash    prevHeaderFields)
+                        (headerFieldBlockNo prevHeaderFields)
+            initAcc = AF.Empty anchor
+        if n == 0 then
+          return initAcc
+        else
+          go (n - 1) prevInfo initAcc
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:140]: Just `n`

File: ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs [hunk 32]
```diff
@@ -0,0 +1,326 @@
...
+                     , Maybe (Header TestBlock)
+                     )
+generateChain = sized $ \size -> do
+    genEmptyChain <- frequency [(1, return True), (9, return False)]
+    if genEmptyChain then
+      -- In the @else@ branch, we only generate empty chains anchored /after/
+      -- genesis here, that's why we generate an empty chain anchored at
+      -- genesis here.
+      return (AF.Empty AF.AnchorGenesis, Nothing)
+    else do
```
Comment (mrBliss) [inline ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/VolDB.hs:167]: Remove, because `genChainHelper` can generate empty chains now.

File: ouroboros-consensus/src/Ouroboros/Consensus/Fragment/ValidatedDiff.hs (hunk 24 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Fragment/ValidatedDiff.hs:61]: Can't, it's `l`, not `LedgerState blk`.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs (hunk 25 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:322]: Let's do that in a separate PR.

File: ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs (hunk 28 above)
Comment (mrBliss) [reply ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs:590]: The checks pass with the two bangs in place, but they fail when I remove them. I'll leave it as is for this PR, but address #2341 separately in a follow-up PR (with proper strict types we don't need to think about bangs anymore).

**Comment by mrBliss** [comment]
bors merge

**Comment by mrBliss** [comment]
bors merge
